commit b366d8435ba3ad133829abf24d07c83852bfeab9
Author: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date:   Thu Apr 19 09:33:30 2018 +0100

    v4l2_m2m: Request selection correctly at appropriate times.
    
    On format changed the selection denoting the image crop
    may well change. Reread the selection on format changed and
    update width/height appropriately.
    
    When setting the selection specify the target.

diff --git a/libavcodec/v4l2_context.c b/libavcodec/v4l2_context.c
index efcb042..aec8ce3 100644
--- a/libavcodec/v4l2_context.c
+++ b/libavcodec/v4l2_context.c
@@ -146,6 +146,7 @@ static int v4l2_handle_event(V4L2Context *ctx)
     struct v4l2_format out_fmt = s->output.format;
     struct v4l2_event evt = { 0 };
     int full_reinit, reinit, ret;
+    struct v4l2_selection selection = {0};
 
     ret = ioctl(s->fd, VIDIOC_DQEVENT, &evt);
     if (ret < 0) {
@@ -178,6 +179,14 @@ static int v4l2_handle_event(V4L2Context *ctx)
     if (reinit) {
         s->capture.height = v4l2_get_height(&cap_fmt);
         s->capture.width = v4l2_get_width(&cap_fmt);
+
+        selection.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+        selection.target = V4L2_SEL_TGT_COMPOSE;
+        ret = ioctl(s->fd, VIDIOC_G_SELECTION, &selection);
+        if (ret) {
+            av_log(logger(ctx), AV_LOG_DEBUG, "crop output %dx%d\n", selection.r.width, selection.r.height);
+            /* update the size of the resulting frame */
+        }
     }
 
     if (full_reinit || reinit)
@@ -197,6 +206,11 @@ static int v4l2_handle_event(V4L2Context *ctx)
         if (ret < 0)
             av_log(logger(ctx), AV_LOG_WARNING, "update avcodec height and width\n");
 
+        if (selection.r.width && selection.r.height) {
+            s->avctx->width  = selection.r.width;
+            s->avctx->height = selection.r.height;
+        }
+
         ret = ff_v4l2_m2m_codec_reinit(s);
         if (ret) {
             av_log(logger(ctx), AV_LOG_ERROR, "v4l2_m2m_codec_reinit\n");
diff --git a/libavcodec/v4l2_m2m_dec.c b/libavcodec/v4l2_m2m_dec.c
index 710e40e..81f727a 100644
--- a/libavcodec/v4l2_m2m_dec.c
+++ b/libavcodec/v4l2_m2m_dec.c
@@ -38,7 +38,7 @@ static int v4l2_try_start(AVCodecContext *avctx)
     V4L2m2mContext *s = ((V4L2m2mPriv*)avctx->priv_data)->context;
     V4L2Context *const capture = &s->capture;
     V4L2Context *const output = &s->output;
-    struct v4l2_selection selection;
+    struct v4l2_selection selection = {0};
     int ret;
 
     /* 1. start the output process */
@@ -67,8 +67,11 @@ static int v4l2_try_start(AVCodecContext *avctx)
 
     /* 3. set the crop parameters */
     selection.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
-    selection.r.height = avctx->coded_height;
-    selection.r.width = avctx->coded_width;
+    selection.r.height = avctx->height;
+    selection.r.width = avctx->width;
+    selection.r.left = 0;
+    selection.r.top = 0;
+    selection.target = V4L2_SEL_TGT_COMPOSE;
     ret = ioctl(s->fd, VIDIOC_S_SELECTION, &selection);
     if (!ret) {
         ret = ioctl(s->fd, VIDIOC_G_SELECTION, &selection);

